OIP042 Edit Support, ElasticSearch scaling, Dockerization, and Documentation

OstlerDev opened the pull request:
# Changelog
## Added
- OIP 042 Edit Functionality
- Edit API Endpoints
- Dockerfile to build Binaries
- Dockerfile to build deployable Image
- Block Reorganization handling/support
- Added CHANGELOG
- Added proper `meta.time` and `edit.timestamp` field mapping for the `oip042_edit` index to allow viewing edits on a timeline in Kibana

## Changed
- Rewrote README
- Increase Index Shard count for `oip042_artifacts` to `3` (2 -> 3)
- Increase Index Shard count for `multiparts` to `3` (2 -> 3)
- Increase Index Shard count for `blocks` to `5` (2 -> 5)
- Increase Index Shard count for `transactions` to `6` (2 -> 6)
- Decreased all other indexes to only have a single shard (2 -> 1)
- Disabled Shard Replicas
- Modified Multipart processing to only occur after `initialSync` is complete
- Changed Default RAM for ElasticSearch to 1/4th of the total system RAM
- Changed the Bulk Indexer to only write 10MB of data at a time instead of 80MB
- Filter Blacklisted Records from API Results
- Docker Healthcheck command now uses `/oip/sync/status` to check health
- Push minimal update for marking a Record as no longer the latest during Edit processing
- Disabled searching in `meta.tx` for the `oip042_edit` index (for better performance, as it was not being used)
- Removed some unnecessary checks of length 
- Tidied up Error logging in the Edit processing code
- Increase Edit query size to 10000

## Fixed
- Use `mainnet` instead of `livenet` as default network name
- Added missing SpatialData index field
- Multipart logic contains a Memory Leak when completing large quantities of records
- EventBus occasionally locks up causing all new Blocks and Records to cease being processed
- Bug in block reorganization handling that could cause the EventBus to lockup if a Datastore happened during a reorganization
- Removed race condition in Incoming Block gap connections

bitspill commented on sync/blockBuffer.go:
```diff
@@ -22,6 +22,16 @@ func (bb *blockBuffer) Push(b *datastore.BlockData) {
 	}
 }
 
+func (bb *blockBuffer) QuickPopFront() *datastore.BlockData {
```
What does QuickPopFront do differently than PopFront and why was it needed?

It looks as though a different block than the front will be returned

bitspill replied:
I ended up coming back around to this and making a quick test to confirm the two functions do return different results

```go
func TestQuick(t *testing.T) {
	bb := blockBuffer{}
	bb.Push(&datastore.BlockData{SecSinceLastBlock: 1})
	bb.Push(&datastore.BlockData{SecSinceLastBlock: 2})
	a := bb.PopFront().SecSinceLastBlock
	fmt.Println(a)

	bb = blockBuffer{}
	bb.Push(&datastore.BlockData{SecSinceLastBlock: 1})
	bb.Push(&datastore.BlockData{SecSinceLastBlock: 2})
	b := bb.QuickPopFront().SecSinceLastBlock
	fmt.Println(b)

	if a != b {
		t.Errorf("Value mismatch. %d != %d", a, b)
	}
}
```

```
=== RUN   TestQuick
2
1
--- FAIL: TestQuick (0.00s)
    blockBuffer_test.go:131: Value mismatch 2 != 1
FAIL
```

OstlerDev replied:
When I was implementing the reorganization code, I was trying to find a way to speed up the return of the new head after popping a block off, which is why this code exists.

After discussing with you on RocketChat, we decided to swap to using only "PopFront" and logging the popped block instead of the new `recentBlocks` head. I will remove the now unused `QuickPopFront` function :)

bitspill commented:
Within edit applications (specifically `queryIncompleteEdits`) I don't see a search size specified so it'll return the ES default of only 10 results and it also appears to only do 1 page of results per datastore commit rather than covering all relevant edits similar to multiparts with the `moreMultiparts` repetition


Along with that I didn't see a mechanism to mark invalid edits and believe that leaves open an opportunity for attack

For example I believe sending 10 edits with non-existent originalTxid would cause all future edits to be ignored

During application the original txid is found to not exist so the edit is ignored however since they comprise the entire set of pending edits for this iteration no edits are processed. When the next datastore commit takes place incomplete edits are queried and because they must be processed oldest-first the entire queue is filled with the same invalid edits again and the process repeats indefinitely stalling the process

OstlerDev commented on datastore/mappings/oip042_artifact.json:
```diff
@@ -1,10 +1,11 @@
   "mappings": {
     "_doc": {
-      "dynamic": "strict",
+      "dynamic": "true",
```
I will look more into this, but I believe we should be ok with having only `details` be dynamic

OstlerDev replied:
After testing with it set to `strict` it seems to work just fine with MLG's Records, so I will go ahead and make this change as well.

OstlerDev commented on modules/oip042/api.go:
```diff
@@ -48,3 +56,144 @@ func handleLatest(w http.ResponseWriter, r *http.Request) {
+		query = elastic.NewBoolQuery().Must(
+			elastic.NewTermQuery("meta.blacklist.blacklisted", false),
+			elastic.NewTermQuery("meta.deactivated", false),
+			elastic.NewTermQuery("meta.originalTxid", originalTxid),
```
You are absolutely correct, I will go ahead and implement this change :)

OstlerDev replied:
Actually, since this API call is grabbing a specific edit, so it doesn't need it's txid swapped.

The `originalTxid` variable holds the initial txid of the first Record published, not of the edit :)

OstlerDev commented on modules/oip042/artifact.go:
```diff
@@ -60,19 +61,30 @@ func on42JsonPublishArtifact(artifact jsoniter.Any, tx *datastore.TransactionDat
+		Blacklist:     Blacklist{Blacklisted: bl, Filter: label},
+		Deactivated:   false,
+		Latest:        true,
+		OriginalTxid:  tx.Transaction.Txid, //todo: this will need to change as part of edit implementation
```
The todo is not relevant for OriginalTxid, but that comment would be valid for the "previousEdits" array that is currently not being used (but should!), I will go ahead and remove it.

OstlerDev commented on modules/oip042/edit.go:
```diff
@@ -0,0 +1,304 @@
+	}
+
+	// Grab the patch string
+	spatchString := string(editRecord.Patch)
```
It should just be `patchString`, I will go ahead and make that change. I believe at one point the variable was just called `spatch` for "string patch", but was later renamed incorrectly

OstlerDev commented on modules/oip042/edit.go:
```diff
@@ -0,0 +1,304 @@
+	}
+
+	// Check if we have remove operations
+	if len(squashedPatch.Remove) > 0 {
```
👍 good to know, I will go ahead and clean this up and remove the unnecessary code

OstlerDev commented on modules/oip042/edit.go:
```diff
@@ -0,0 +1,304 @@
+			return nil, err
+		}
+		
+		return artifactRecord, nil
```
Sounds good, I will replace it with that

OstlerDev commented on modules/oip042/edit.go:
```diff
@@ -0,0 +1,304 @@
+	artifactRecord.Meta.Latest = false
+
+	// Run updates to set "latest" to false on the previously latest Record
+	cu := datastore.Client().Update().Index(datastore.Index(oip042ArtifactIndex)).Type("_doc").Id(artifactRecord.Meta.Txid).Doc(artifactRecord).Refresh("false")
```
Great plan, I will implement that :)

OstlerDev replied:
👍 I will change this and push up an update

OstlerDev commented:
> For example I believe sending 10 edits with non-existent originalTxid would cause all future edits to be ignored

@bitspill You are absolutely right and this is actually something we ran into in our Dev environment. For now I have set an Edit query size of 10,000, as that should mitigate most of the potential attack vector for now. I do agree with you that we do need to implement marking Edits as "broken" or "stale", but I am thinking that it might have to come in a later PR. I will open an issue here on GitHub so that we remember we still need to do this piece of work 😄

OstlerDev commented on datastore/mappings/oip042_edit.json:
```diff
@@ -0,0 +1,289 @@
+            "time": {
+              "type": "long"
+            },
+            "tx": {
```
👍 Disabled!

OstlerDev commented:
The latest commit I have pushed up to this PR reflects the latest MLG version. It contains the following additional (requested) changes.

## Changelog
### Added
- Added proper `meta.time` and `edit.timestamp` field mapping for the `oip042_edit` index to allow viewing edits on a timeline in Kibana

### Changed
- Push minimal update for marking a Record as no longer the latest during Edit processing
- Formatted the code using `go fmt`
- Disabled searching in `meta.tx` for the `oip042_edit` index (for better performance, as it was not being used)
- Removed some unnecessary checks of length 
- Tidied up Error logging in the Edit processing code
- Increase Edit query size to 10000

### Fixed
- Bug in block reorganization handling that could cause the EventBus to lockup if a Datastore happened during a reorganization
- Removed race condition in Incoming Block gap connections

cchrysostom commented:
Let's address the two issues caught by GolangCI.

OstlerDev commented:
> Let's address the two issues caught by GolangCI.

The two issues that were being reported by GolangCI have now been addressed 😊 All CI checks are now passing.

OstlerDev commented:
No problem at all, @bitspill it is now ready for your review :)

OstlerDev commented on modules/oip042/edit.go:
```diff
@@ -145,90 +148,106 @@ func queryArtifact(txid string) (*elasticOip042Artifact, error) {
+
+	signatureOk, err = flo.CheckSignature(floAddress, signature, preImage)
+	if !signatureOk {
+		return fmt.Errorf("Editted Record has invalid Signature! Address: %v | Preimage: %v | Signature: %v | Error: %v", floAddress, preImage, signature, err)
```
You are absolutely correct, for now I am going to leave it though, since it is just a text issue hat should only occur rarely 😃

OstlerDev commented:
MLG is currently using this same code on our Production servers with no issue, I have performed multiple synchronizations on both `mainnet` and `testnet`  🙂

I am going to go ahead and merge this in 👍
